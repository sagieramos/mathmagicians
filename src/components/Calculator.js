import React, { useEffect, useReducer } from 'react';
import { v4 as uuidv4 } from 'uuid';
import CalculatorButton from './CalculatorButton';
import calculate from '../logic/calculate';
import QuoteComponent from './QuoteComponent';
import { reducer, initialState } from './reducer';
import './Calculator.scss';

const api = {
  Key: 'qq3TQU5AD1tRAqUt3kq5Eg==KTG05ywHFyq0JPk1',
  url: 'https://api.api-ninjas.com/v1/quotes?category=love',
};

const Calculator = () => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const handleBtnClick = (value) => {
    const newCalcState = calculate(state.calcState, value);
    dispatch({ type: 'SET_CALC_STATE', payload: newCalcState });
    const { next, total, operation } = newCalcState;
    dispatch({ type: 'SET_INPUT', payload: next || total || operation || '' });
  };

  useEffect(() => {
    (async () => {
      try {
        const response = await fetch(api.url, {
          headers: {
            'X-Api-Key': api.Key,
          },
        });

        if (!response.ok) {
          throw new Error('Network response was not ok');
        }

        const data = await response.json();
        dispatch({
          type: 'SET_QUOTE_DATA',
          payload: { quote: data[0], loading: false, error: null },
        });
      } catch (error) {
        dispatch({
          type: 'SET_QUOTE_DATA',
          payload: { quote: { quote: '', author: '', category: '' }, loading: false, error: error.message },
        });
      }
    })();
  }, []);

  let quoteContent;
  if (state.quoteData.loading) {
    quoteContent = <p>Loading...</p>;
  } else if (state.quoteData.error) {
    quoteContent = (
      <p>
        Error:
        {' '}
        {state.quoteData.error}
      </p>
    );
  } else {
    const { quote } = state.quoteData;
    quoteContent = <QuoteComponent quote={quote} />;
  }

  return (
    <div id="calc">
      <div className="display-calc">
        <input type="text" value={state.input} readOnly />
      </div>
      <div className="btn-calc">
        {[
          ['AC', '+/-', '%', 'รท'],
          ['7', '8', '9', 'x'],
          ['4', '5', '6', '-'],
          ['1', '2', '3', '+'],
          ['0', '.', '='],
        ].map((row) => (
          <div key={uuidv4()} className="row-calc">
            {row.map((value) => (
              <CalculatorButton
                key={value}
                value={value}
                onClick={() => handleBtnClick(value)}
              />
            ))}
          </div>
        ))}
      </div>
      <div key={uuidv4()} className="quote">
        {quoteContent}
      </div>
    </div>
  );
};

export default Calculator;
