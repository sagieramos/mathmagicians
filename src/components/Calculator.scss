@import url("https://fonts.googleapis.com/css2?family=Poppins:wght@400;500;600;700;800&display=swap");

$breakpoint-mobile: 412px;
$breakpoint-mid: 800px;
$background-color: #f4f4f4;

@mixin bakgrd($bg: inherit, $color: #fff) {
  width: 100%;
  border: none;
  background: $bg;
  color: $color;
  font-size: 1em;
  font-weight: 600;
}

#calc-wrapper {
  @media screen and (max-width: $breakpoint-mid) {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;

    h2 {
      align-self: flex-start;
      padding: 0 32px;
    }
  }

  @media screen and (min-width: $breakpoint-mid) {
    display: flex;
    align-items: first baseline;
    justify-content: space-between;
    gap: 32px;
    padding: 0 64px;

    h2 {
      align-self: flex-start;
      padding: 0 32px;
      word-spacing: 24px;
      letter-spacing: 12px;
    }
  }

  #calc {
    font-family: "Poppins", sans-serif;
    max-width: 412px;
    min-width: 252px;

    * {
      box-sizing: border-box;
      margin: 0;
      padding: 0;
    }

    @media screen and (max-width: $breakpoint-mobile) {
      width: 100%;
    }

    @media screen and (min-width: $breakpoint-mobile) {
      width: 90%;
    }

    .display-calc {
      background: #919297;

      input {
        @include bakgrd();

        padding: 16px;
        text-align: end;
        cursor: auto;
      }
    }

    .btn-calc {
      .row-calc {
        width: 100%;
        display: flex;

        button {
          flex: 1;
          text-align: center;
          padding: 16px;
          border: 1px solid #cccccc83;
          background-color: #f5f5f5;
          cursor: pointer;
          transition: background-color 0.3s ease;

          &:last-child {
            background: #f99245;
          }

          &:active {
            background-color: #919297;
            color: #fff;
          }
        }
      }

      :last-child {
        :first-child {
          flex: 2;
          padding: 16px 33px;
        }
      }
    }
  }
}
