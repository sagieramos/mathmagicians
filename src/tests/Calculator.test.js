import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Calculator from '../components/Calculator';

test('should render the calculator input correctly', () => {
  render(<Calculator />);
  const input = screen.getByRole('textbox');
  expect(input).toBeInTheDocument();
});
test('should render the calculator correctly', () => {
  render(<Calculator />);
  const calculator = screen.getByTestId('calc-wrapper');
  expect(calculator).toMatchSnapshot();
});
test('should update the calculator input when the user presses a number button', () => {
  render(<Calculator />);
  const input = screen.getByRole('textbox');
  fireEvent.click(screen.getByText('1'));
  expect(input.value).toBe('1');
});
