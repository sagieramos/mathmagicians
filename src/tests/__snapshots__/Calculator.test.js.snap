// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the calculator correctly 1`] = `
<div
  data-testid="calc-wrapper"
  id="calc-wrapper"
>
  <h2>
    Let's do some math!
  </h2>
  <div
    id="calc"
  >
    <div
      class="display-calc"
    >
      <input
        readonly=""
        type="text"
        value=""
      />
    </div>
    <div
      class="btn-calc"
    >
      <div
        class="row-calc"
      >
        <button
          type="button"
        >
          AC
        </button>
        <button
          type="button"
        >
          +/-
        </button>
        <button
          type="button"
        >
          %
        </button>
        <button
          type="button"
        >
          รท
        </button>
      </div>
      <div
        class="row-calc"
      >
        <button
          type="button"
        >
          7
        </button>
        <button
          type="button"
        >
          8
        </button>
        <button
          type="button"
        >
          9
        </button>
        <button
          type="button"
        >
          x
        </button>
      </div>
      <div
        class="row-calc"
      >
        <button
          type="button"
        >
          4
        </button>
        <button
          type="button"
        >
          5
        </button>
        <button
          type="button"
        >
          6
        </button>
        <button
          type="button"
        >
          -
        </button>
      </div>
      <div
        class="row-calc"
      >
        <button
          type="button"
        >
          1
        </button>
        <button
          type="button"
        >
          2
        </button>
        <button
          type="button"
        >
          3
        </button>
        <button
          type="button"
        >
          +
        </button>
      </div>
      <div
        class="row-calc"
      >
        <button
          type="button"
        >
          0
        </button>
        <button
          type="button"
        >
          .
        </button>
        <button
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>
</div>
`;
